### PR TITLE
feat: kratos cookie flow 2

### PR DESCRIPTION
**Pull request by ntheile**
This enables the kratos cookie login flow for browser login. See reverted PR for more details https://github.com/GaloyMoney/galoy/pull/2149

### CODE REVIEW

File: dev/ory/oathkeeper_rules.yaml
Comment (ntheile) [inline dev/ory/oathkeeper_rules.yaml:31]: @krtk6160 

I had to change this because there are a few routes that were missed.  I added the matcher to exclude the `auth/clearCookies` endpoint. if this is not set then it has two matches and won't resolve. this endpoint needs to allow a user to pass expired or revoked cookies so the server can clear them for the browser and avoid login issues. the matcher still allows all wildcard endpoints, such as `graphql`.

File: dev/ory/oathkeeper_rules.yaml
Comment (ntheile) [inline dev/ory/oathkeeper_rules.yaml:47]: @nicolasburtey adding this to use kratos instead of our custom `validatetoken` endpoint. I needed to change the `subject_from` field because it was returning blank when we switched from our graphql api `auth/validatetoken` to the kratos `whoami` endpoint

**Comment by ntheile** [comment]
@openoms This code should enable cookie auth. I will test in my local dev charts after this PR is merged https://github.com/GaloyMoney/charts/pull/2474

File: dev/ory/oathkeeper_rules.yaml
Comment (ntheile) [reply dev/ory/oathkeeper_rules.yaml:47]: It appears this makes the github actions e2e test fail with this error 

```
header › getting a kratos header when passing a legacy JWT in the header
AxiosError: Request failed with status code 401
```
Are the legacy jwt's fully depreciated yet?

Comment (ntheile) [reply dev/ory/oathkeeper_rules.yaml:47]: going to revert this back because the admin panel is still using the jwts

**Comment by ntheile** [comment]
not sure why e2e tests keep failing with the error. any ideas @krtk6160 or @nicolasburtey ?

```
error createAccountWithPhoneIdentifier: UnknownRepositoryError: Plan executor error during findAndModify :: caused by :: E11000 duplicate key error collection: galoy.users index: phone_1 dup key: { phone: "+1415388495" }
```

**Comment by krtk6160** [comment]
> not sure why e2e tests keep failing with the error. any ideas @krtk6160 or @nicolasburtey ?
> 
> ```
> error createAccountWithPhoneIdentifier: UnknownRepositoryError: 
> Plan executor error during findAndModify :: caused by :: E11000 duplicate 
> key error collection: galoy.users index: phone_1 dup key: { phone: "+1415388495" }
> ```

That is not really erroring. We see those messages in the logs even when the e2e tests pass.

File: dev/ory/oathkeeper_rules.yaml
Comment (nicolasburtey) [inline dev/ory/oathkeeper_rules.yaml:98]: name is not great imo. is there something more descriptive we can have?

it's not great because we don't know from the name what route the "non other" represents

File: src/domain/authentication/index.types.d.ts
Comment (nicolasburtey) [inline src/domain/authentication/index.types.d.ts:9]: domain should not have `Kratos` @bodymindarts ?

File: src/servers/middlewares/auth-router.ts
Comment (nicolasburtey) [inline src/servers/middlewares/auth-router.ts:128]: could we not have a big try catch?

**Review by nicolasburtey — COMMENTED** [review]
some comments. I don't have a lot of understanding on the nuances of cookies so someone else should review and approve

File: src/app/auth/login.ts
Comment (dolcalmi) [inline src/app/auth/login.ts:98]: maybe better to use Promise.All

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:216]: shouldnt we return here?

File: src/services/kratos/auth-phone-no-password.ts
Comment (dolcalmi) [inline src/services/kratos/auth-phone-no-password.ts:110]: use `err.message || err`. This applies to all methods in this service

File: test/e2e/servers/ory/kratos.spec.ts
Comment (dolcalmi) [inline test/e2e/servers/ory/kratos.spec.ts:652]: this create a conditional test.. it should be independent of `login with cookie`

File: test/e2e/servers/ory/kratos.spec.ts
Comment (dolcalmi) [inline test/e2e/servers/ory/kratos.spec.ts:670]: this is a conditional test. please rebase to get the lint warning

**Review by dolcalmi — CHANGES_REQUESTED** [review]

**Review by nicolasburtey — APPROVED** [review]

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:126]: namespaces should include layers structure. so this should be more like `servers.middlewares.authRouter` and please add `fnName` with `login`

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:204]: same here. namespace `servers.middlewares.authRouter` and please add `fnName` with `logout`

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:206]: trycatch comment apply to this method too

File: src/services/kratos/auth-phone-no-password.ts
Comment (dolcalmi) [inline src/services/kratos/auth-phone-no-password.ts:105]: please review all catches in the service... `err.message || err`

File: src/services/kratos/auth-phone-no-password.ts
Comment (dolcalmi) [inline src/services/kratos/auth-phone-no-password.ts:187]: are you sure this will not throw?

**Review by dolcalmi — CHANGES_REQUESTED** [review]

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [reply src/servers/middlewares/auth-router.ts:140]: if Auth is properly build in app/services layer then there is no need to be in try block

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:152]: you need to use `recordExceptionInCurrentSpan`

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:183]: same here: `recordExceptionInCurrentSpan`

File: src/servers/middlewares/auth-router.ts
Comment (dolcalmi) [inline src/servers/middlewares/auth-router.ts:231]: same here `recordExceptionInCurrentSpan`

File: src/services/kratos/auth-phone-no-password.ts
Comment (dolcalmi) [inline src/services/kratos/auth-phone-no-password.ts:87]: are you sure this will not throw?

**Review by dolcalmi — CHANGES_REQUESTED** [review]

**Review by dolcalmi — DISMISSED** [review]

**Review by nicolasburtey — APPROVED** [review]

**Pull request merged**
